File: test/entities/vbkblock_test.cpp
// Copyright (c) 2019-2020 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

#include "veriblock/entities/vbkblock.hpp"
#include "veriblock/blockchain/block_index.hpp"

#include <gtest/gtest.h>

#include "veriblock/arith_uint256.hpp"
#include "veriblock/literals.hpp"

using namespace altintegration;

static const VbkBlock defaultBlock{5000,
                                   2,
                                   "449c60619294546ad825af03"_unhex,
                                   "b0935637860679ddd5"_unhex,
                                   "5ee4fd21082e18686e"_unhex,
                                   "26bbfda7d5e4462ef24ae02d67e47d78"_unhex,
                                   1553699059,
                                   16842752,
                                   1};

static const std::string defaultBlockEncoded =
    "41000013880002449c60619294546ad825af03b0935637860679ddd55ee4fd21082e18686e"
    "26bbfda7d5e4462ef24ae02d67e47d785c9b90f3010100000000000001";

TEST(VbkBlock, Deserialize) {
  const auto vbkblock = ParseHex(defaultBlockEncoded);
  auto stream = ReadStream(vbkblock);
  auto block = VbkBlock::fromVbkEncoding(stream);

  EXPECT_EQ(block.getHeight(), defaultBlock.getHeight());
  EXPECT_EQ(block.getVersion(), defaultBlock.getVersion());
  EXPECT_EQ(block.getPreviousBlock().toHex(),
            defaultBlock.getPreviousBlock().toHex());
  EXPECT_EQ(block.getPreviousKeystone().toHex(),
            defaultBlock.getPreviousKeystone().toHex());
  EXPECT_EQ(block.getSecondPreviousKeystone().toHex(),
            defaultBlock.getSecondPreviousKeystone().toHex());
  EXPECT_EQ(block.getMerkleRoot().toHex(),
            defaultBlock.getMerkleRoot().toHex());
  EXPECT_EQ(block.getBlockTime(), defaultBlock.getBlockTime());
  EXPECT_EQ(block.getDifficulty(), defaultBlock.getDifficulty());
  EXPECT_EQ(block.getNonce(), defaultBlock.getNonce());
}

TEST(VbkBlock, Serialize) {
  WriteStream stream;
  defaultBlock.toVbkEncoding(stream);
  auto vbkBytes = stream.data();
  auto blockEncoded = HexStr(vbkBytes);
  EXPECT_EQ(blockEncoded, defaultBlockEncoded);
}

TEST(VbkBlock, RoundTrip) {
  auto blockEncoded = ParseHex(defaultBlockEncoded);
  auto stream = ReadStream(blockEncoded);
  auto decoded = VbkBlock::fromVbkEncoding(stream);
  EXPECT_EQ(decoded.getVersion(), defaultBlock.getVersion());

  WriteStream outputStream;
  decoded.toVbkEncoding(outputStream);
  auto vbkBytes = outputStream.data();
  auto blockReEncoded = HexStr(vbkBytes);
  EXPECT_EQ(blockReEncoded, defaultBlockEncoded);
}

TEST(VbkBlock, RoundTripNew) {
  auto blockEncoded = ParseHex(defaultBlockEncoded);
  VbkBlock decoded;
  ValidationState state;
  bool ret = Deserialize(blockEncoded, decoded, state);
  ASSERT_TRUE(ret);
  EXPECT_TRUE(state.IsValid());
  EXPECT_EQ(decoded.getVersion(), defaultBlock.getVersion());

  WriteStream outputStream;
  decoded.toVbkEncoding(outputStream);
  auto vbkBytes = outputStream.data();
  auto blockReEncoded = HexStr(vbkBytes);
  EXPECT_EQ(blockReEncoded, defaultBlockEncoded);
}

TEST(VbkBlock, RoundTripWithHash) {
  const auto blockEncoded = defaultBlock.toRawAddHash();
  auto bytes = std::string(blockEncoded.begin(), blockEncoded.end());
  auto decoded = VbkBlock::fromRawAddHash(bytes);
  EXPECT_EQ(decoded, defaultBlock);
}

TEST(VbkBlock, RoundTripBlockIndexWithHash) {
  BlockIndex<VbkBlock> defaultBlockIndex{};
  defaultBlockIndex.setHeader(std::make_shared<VbkBlock>(defaultBlock));
  const auto blockEncoded = defaultBlockIndex.toRawAddHash();
  auto bytes = std::string(blockEncoded.begin(), blockEncoded.end());
  auto decoded = BlockIndex<VbkBlock>::fromRawAddHash(bytes);
  EXPECT_EQ(decoded.getHeader(), defaultBlock);
}

TEST(VbkBlock, getBlockHash_test) {
  VbkBlock block;
  block.setHeight(5000);
  block.setVersion(2);
  block.setPreviousBlock(uint96("94E7DC3E3BE21A96ECCF0FBD"_unhex));
  block.setPreviousKeystone(uint72("F5F62A3331DC995C36"_unhex));
  block.setSecondPreviousKeystone(uint72("B0935637860679DDD5"_unhex));
  block.setMerkleRoot(uint128("DB0F135312B2C27867C9A83EF1B99B98"_unhex));
  block.setTimestamp(1553699987);
  block.setDifficulty(117586646);
  block.setNonce(1924857207);

  EXPECT_EQ(
      ArithUint256::fromLEBytes(block.getHash()),
      ArithUint256::fromHex(
          "00000000000000001f45c91342b8ac0ea7ae4d721be2445dc86ddc3f0e454f60"));
}

TEST(VbkBlock, getId_test) {
  auto atvBytes = ParseHex(defaultBlockEncoded);
  auto stream = ReadStream(atvBytes);
  auto vbkblock = VbkBlock::fromVbkEncoding(stream);

  EXPECT_EQ(vbkblock.getId().toHex(), "cd97599e23096ad42f119b5a");
}

File: bindings/go/entities/vbkblock_test.go
// Copyright (c) 2019-2021 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package entities

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
)

var (
	defaultVbkBlock = VbkBlock{
		5000,
		2,
		parse12Bytes("449c60619294546ad825af03"),
		parse9Bytes("b0935637860679ddd5"),
		parse9Bytes("5ee4fd21082e18686e"),
		parse16Bytes("26bbfda7d5e4462ef24ae02d67e47d78"),
		1553699059,
		16842752,
		1,
	}
	defaultVbkBlockEncoded = "41000013880002449c60619294546ad825af03b0935637860679ddd55ee4fd21082e18686e26bbfda7d5e4462ef24ae02d67e47d785c9b90f3010100000000000001"
)

func parse12Bytes(src string) [12]byte {
	buf := parseHex(src)
	var block [12]byte
	copy(block[:], buf)
	return block
}

func parse9Bytes(src string) [9]byte {
	buf := parseHex(src)
	var block [9]byte
	copy(block[:], buf)
	return block
}

func parse16Bytes(src string) [16]byte {
	buf := parseHex(src)
	var block [16]byte
	copy(block[:], buf)
	return block
}

func TestVbkBlockDeserialize(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	vbkblock := parseHex(defaultVbkBlockEncoded)
	stream := bytes.NewReader(vbkblock)
	block := VbkBlock{}
	assert.NoError(block.FromVbkEncoding(stream))

	assert.Equal(defaultVbkBlock.Height, block.Height)
	assert.Equal(defaultVbkBlock.Version, block.Version)
	assert.Equal(hex.EncodeToString(defaultVbkBlock.PreviousBlock[:]), hex.EncodeToString(block.PreviousBlock[:]))
	assert.Equal(hex.EncodeToString(defaultVbkBlock.PreviousKeystone[:]), hex.EncodeToString(block.PreviousKeystone[:]))
	assert.Equal(hex.EncodeToString(defaultVbkBlock.SecondPreviousKeystone[:]), hex.EncodeToString(block.SecondPreviousKeystone[:]))
	assert.Equal(hex.EncodeToString(defaultVbkBlock.MerkleRoot[:]), hex.EncodeToString(block.MerkleRoot[:]))
	assert.Equal(defaultVbkBlock.Timestamp, block.Timestamp)
	assert.Equal(defaultVbkBlock.Difficulty, block.Difficulty)
	assert.Equal(defaultVbkBlock.Nonce, block.Nonce)
}

func TestVbkBlockSerialize(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	stream := new(bytes.Buffer)
	assert.NoError(defaultVbkBlock.ToVbkEncoding(stream))
	blockEncoded := hex.EncodeToString(stream.Bytes())
	assert.Equal(defaultVbkBlockEncoded, blockEncoded)
}

func TestVbkBlockRoundTrip(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	blockEncoded, err := hex.DecodeString(defaultVbkBlockEncoded)
	assert.NoError(err)
	stream := bytes.NewReader(blockEncoded)
	decoded := VbkBlock{}
	assert.NoError(decoded.FromVbkEncoding(stream))
	assert.Equal(defaultVbkBlock.Version, decoded.Version)

	outputStream := new(bytes.Buffer)
	assert.NoError(decoded.ToVbkEncoding(outputStream))
	assert.Equal(defaultVbkBlockEncoded, hex.EncodeToString(outputStream.Bytes()))
}

func TestVbkBlockGetBlockHash(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	block := VbkBlock{}
	block.Height = 5000
	block.Version = 2
	copy(block.PreviousBlock[:], parseHex("94E7DC3E3BE21A96ECCF0FBD"))
	copy(block.PreviousKeystone[:], parseHex("F5F62A3331DC995C36"))
	copy(block.SecondPreviousKeystone[:], parseHex("B0935637860679DDD5"))
	copy(block.MerkleRoot[:], parseHex("DB0F135312B2C27867C9A83EF1B99B98"))
	block.Timestamp = 1553699987
	block.Difficulty = 117586646
	block.Nonce = 1924857207

	hash := block.GetHash()
	assert.Equal("1f45c91342b8ac0ea7ae4d721be2445dc86ddc3f0e454f60", hex.EncodeToString(hash[:]))
}

func TestVbkBlockGetID(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	atvBytes, err := hex.DecodeString(defaultVbkBlockEncoded)
	assert.NoError(err)
	stream := bytes.NewReader(atvBytes)
	vbkblock := VbkBlock{}
	assert.NoError(vbkblock.FromVbkEncoding(stream))
	id := vbkblock.GetID()
	assert.Equal("cd97599e23096ad42f119b5a", hex.EncodeToString(id[:]))
}

func TestVbkToJSON(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	defaultJSON := `{"difficulty":16842752,"hash":"2ba076219b4ff7ed36512275cd97599e23096ad42f119b5a","height":5000,"id":"cd97599e23096ad42f119b5a","merkleRoot":"26bbfda7d5e4462ef24ae02d67e47d78","nonce":1,"previousBlock":"449c60619294546ad825af03","previousKeystone":"b0935637860679ddd5","secondPreviousKeystone":"5ee4fd21082e18686e","timestamp":1553699059,"version":2}`
	jsonmap, err := defaultVbkBlock.ToJSON()
	assert.NoError(err)

	res, err := json.Marshal(jsonmap)
	assert.NoError(err)
	assert.Equal(defaultJSON, string(res))
}
